src/support.cc

```diff
@@ -528,135 +528,264 @@ bool digest_message(const char* alg, const byte* message, int message_len,
+  int tag_len = 0;
+  byte tag[16];
+  int aad_len = 0;
+  byte* aad = nullptr;
+  bool ret = true;
+
+  if(!(ctx = EVP_CIPHER_CTX_new())) {
+    return false;
+  }
+
+  if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr)) {
+    ret = false;
+    goto done;
+  }
+
+  // set IV length
+  if(1 != EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, blk_size, nullptr)) {
+    ret = false;
+    goto done;
+  }
+  if(1 != EVP_EncryptInit_ex(ctx, nullptr, nullptr, key, iv)) {
+    ret = false;
+    goto done;
+  }
+
+  memcpy(out, iv, blk_size);
+
+  if(1 != EVP_EncryptUpdate(ctx, nullptr, &len, aad, aad_len)) {
+    ret = false;
+    goto done;
+  }
+  if(1 != EVP_EncryptUpdate(ctx, out + blk_size, &len, in, in_len)) {
+    ret = false;
+    goto done;
+  }
+  ciphertext_len = len + blk_size;
+
+  // Finalize
+  if(1 != EVP_EncryptFinal_ex(ctx, out + len, &len)) {
+    ret = false;
+    goto done;
+  }
+  ciphertext_len += len;
+
+  tag_len = EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, blk_size, tag);
+  if(tag_len <= 0) {
+    ret = false;
+    goto done;
+  }
+
+  // append tag
+  memcpy(out + ciphertext_len, tag, blk_size);
+  *out_size = ciphertext_len + blk_size;
+
+done:
+  if (ctx != nullptr) {
+    EVP_CIPHER_CTX_free(ctx);
+    ctx = nullptr;
+  }
+  return ret;
+}
+
+// We use 128 bit tag
+bool aes_256_gcm_decrypt(byte* in, int in_len, byte *key,
+            byte *out, int* out_size) {
+  EVP_CIPHER_CTX *ctx = nullptr;
+  int blk_size =  cipher_block_byte_size("aes-256");
+  int key_size =  cipher_key_byte_size("aes-256");
+  byte* iv = in;
+  bool ret = true;
+  byte* tag = in + in_len - blk_size;
+  int aad_len = 0;
+  byte* aad = nullptr;
+  int len;
+  int plaintext_len;
+  int stream_len = in_len - 2 * blk_size;
+  int err = 0;
+
+  if(!(ctx = EVP_CIPHER_CTX_new())) {
+    return false;
+  }
+  if(!EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr)) {
+    ret = false;
+    goto done;
+  }
+  if(!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, blk_size, nullptr)) {
+    ret = false;
+    goto done;
+  }
+  if(!EVP_DecryptInit_ex(ctx, nullptr, nullptr, key, iv)) {
+    ret = false;
+    goto done;
+  }
+
+  if(!EVP_DecryptUpdate(ctx, nullptr, &len, aad, aad_len)) {
+    ret = false;
+    goto done;
+  }
+  if(!EVP_DecryptUpdate(ctx, out, &len, in + blk_size, stream_len)) {
+    ret = false;
+    goto done;
+  }
+  plaintext_len = len;
+
+  if(!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, blk_size, tag)) {
+    ret = false;
+    goto done;
+  }
+
+  // Finalize
+  err = EVP_DecryptFinal_ex(ctx, in + in_len - blk_size, &len);
+  if (err <= 0) {
+    ret = false;
+    goto done;
+  }
+
+  *out_size = plaintext_len;
+
+done:
+  if (ctx != nullptr)
+    EVP_CIPHER_CTX_free(ctx);
+  return ret;
+}
+
+bool authenticated_encrypt(const char* alg_name, byte* in, int in_len, byte *key,
+            byte *iv, byte *out, int* out_size) {
+
+  if (strcmp(alg_name, "aes-256-cbc-hmac-sha256") == 0) {
```

Comment:
As this is new code, I'm making this comment. You may choose to do nothing right now.

Qs. Is `alg_name` null-terminated?

We are advised to use `strncmp()` for safe-comparison and to avoid comparing run-away string lengths.

I see that existing code uses `strcmp()` liberally, so perhaps, we can shelf this idea for now. In a future rev, I'd be happy to work with Ye & you to introduce `STRCMP()` and `STRNCMP()` as boolean comparison macros and tighten up this code.